### PR TITLE
Adding buffer geometry utilities for merging.

### PR DESCRIPTION
**Pull request by kylegmaxwell**
These are some buffer geometry utilities that I wrote for merging geometry and related operations. I am submitting this after discussion on {mrdoob}/three.js#8162 with @WestLangley 
https://github.com/mrdoob/three.js/pull/8162#issuecomment-238746823
The main idea is that to merge buffer geometry requires creating a new geometry, hence the functions are static on the BufferGeometry class. I also created a jsfiddle as an example.
https://jsfiddle.net/kylu/c1rt9dnb/

### CODE REVIEW

**Comment by mrdoob** [comment]
I'm thinking of creating a `InstancedBufferGeometry`. `splitVertices` would then become something like ```if ( geometry.isInstancedBufferGeometry ) geometry2 = geometry.toBufferGeometry();```, right?

**Comment by mrdoob** [comment]
Which, actually, is code that we already have:

https://github.com/mrdoob/three.js/blob/dev/src/core/BufferGeometry.js#L842-L885

**Comment by kylegmaxwell** [comment]
Oh, good call. I just updated the PR with another commit to use the toNonIndexed function.

File: src/core/BufferGeometry.js [hunk 1]
```diff
@@ -1148,15 +1148,10 @@ BufferGeometry.merge = function( geometries ) {
...
 		if ( geom2.index ) {
 
-			geom2 = BufferGeometry.splitVertices( geom2 );
+			geom2 = geom2.toNonIndexed();
 			geometries[ m ] = geom2;
 
 		}
-		if ( ! geom2.attributes.normal ) {
-
-			// geometries[ m ].geometry = computeNormals( geom2 );
```
Comment (kylegmaxwell) [inline src/core/BufferGeometry.js]: Not sure if anyone is interested, but I also have a function that computes normal vectors that are sharp for corners, and smooth for flat regions. It uses an angle threshold to determine whether the surface is smooth or sharp. I left it out of this pull request.

File: src/core/BufferGeometry.js [hunk 2]
```diff
@@ -642,9 +642,49 @@ Object.assign( BufferGeometry.prototype, EventDispatcher.prototype, {
 
 	}(),
 
+	/**
+	 * Allows flat shading of faces on non-indexed buffer geometry
+	 */
 	computeFaceNormals: function () {
```
Comment (WestLangley) [inline src/core/BufferGeometry.js]: Doesn't `computeVertexNormals()` do flat shading for non-indexed geometries?

File: src/core/BufferGeometry.js [hunk 3]
```diff
@@ -642,9 +642,49 @@ Object.assign( BufferGeometry.prototype, EventDispatcher.prototype, {
...
+			pos[ iC + 1 ] - pos[ iB + 1 ],
+			pos[ iC + 2 ] - pos[ iB + 2 ] );
+			ab.set( pos[ iA ] - pos[ iB ],
+			pos[ iA + 1 ] - pos[ iB + 1 ],
+			pos[ iA + 2 ] - pos[ iB + 2 ] );
+			cb.cross( ab );
+			cb.normalize();
+
+			var face = [ iA, iB, iC ];
+			// For each vertex on the face
```
Comment (WestLangley) [inline src/core/BufferGeometry.js]: I would not instantiate an array in this case just for the purpose of being able to loop over it. Just assign the 9 values.

In any event, I am not sure this method is needed.

File: src/core/BufferGeometry.js [hunk 4]
```diff
@@ -811,6 +851,60 @@ Object.assign( BufferGeometry.prototype, EventDispatcher.prototype, {
 
 	},
 
+	/**
+	 * Checks for duplicate vertices with hashmap.
+	 * Faces with duplicate vertices are rewired to point to the first instance.
+	 * Unused vertices are not removed.
+	 * @param  {THREE.BufferGeometry} geom Geometry to merge
+	 */
+	mergeVertices: function( geom ) {
```
Comment (WestLangley) [inline src/core/BufferGeometry.js]: `this` is not referenced in this method.

Can you please explain in more detail what this method does -- and how it is intended to be used?

File: src/core/BufferGeometry.js (hunk 2 above)
Comment (kylegmaxwell) [inline src/core/BufferGeometry.js]: Oh, I didn't realize that. I will add a comment to the function and remove this.

File: src/core/BufferGeometry.js (hunk 3 above)
Comment (kylegmaxwell) [inline src/core/BufferGeometry.js]: right, I have replaced this with a call to the existing function.

File: src/core/BufferGeometry.js (hunk 4 above)
Comment (kylegmaxwell) [inline src/core/BufferGeometry.js]: This function was for my computeCuspNormals function that I also wrote (but did not include), I will remove this here and file a separate PR later.

**Comment by kylegmaxwell** [comment]
ok, the third commit should address the comments so far, and now the PR is much more concise.

File: src/core/BufferGeometry.js [hunk 5]
```diff
@@ -1036,4 +1045,67 @@ Object.assign( BufferGeometry.prototype, EventDispatcher.prototype, {
...
+		if ( geom1.attributes[ key ].array.constructor !== Float32Array ) continue;
+
+		var data = [];
+		// for each geometry
+		for ( m = 0; m < geometries.length; m ++ ) {
+
+			geom2 = geometries[ m ];
+			if ( ! geom2.attributes[ key ] ) {
+
+				throw new FluxGeometryError( 'Mismatched geometry attributes: ' + key );
```
Comment (WestLangley) [inline src/core/BufferGeometry.js]: `FluxGeometryError`?

**Comment by WestLangley** [comment]
So the restriction is geometries 2, 3, ..., N must have all the attributes of the first geometry in the array -- and any additional attributes in geometries 2, 3, ..., N are ignored?

**Comment by kylegmaxwell** [comment]
Yes, I can supply a default value instead, but it seemed more straightforward to drop the inconsistent attributes. Then the user can set them ahead of time with the appropriate defaults on a case by case basis.

File: src/core/BufferGeometry.js (hunk 5 above)
Comment (kylegmaxwell) [inline src/core/BufferGeometry.js]: Oops, that was a vestige of the old code base where this came from. Fixed now.

**Comment by WestLangley** [comment]
So if the first geometry is a special case, then would it make more sense to start with a geometry?

    bufferGeometry.merge( geometries ); // geometries can be a single geometry or an array

In other words, avoid the static method approach.

**Comment by kylegmaxwell** [comment]
Yes, that's a good idea. Your suggestion to make it a member function seems to fit better with the current design of the class.

Though, this reminds me, there is already a merge function, which seems to merge two buffer geometries, but assumes that the buffers have already been allocated. I must say, I found that counter intuitive at first, coming from using the merge function on Geometry, I expected similar behavior, though I understand buffer geometry has different constraints.

Anyways, I renamed my function join, so as to not conflict with merge.

**Comment by alvaromongon** [comment]
This code seems to work as expected, will it be merged?

**Pull request closed without merging**